addons/website_portal_sale/models/sale_order.py

```diff
@@ -56,6 +75,34 @@ def _notification_recipients(self, message, groups):
 
         return groups
 
+    @api.model_cr_context
+    def _init_column(self, column_name):
+        """ Initialize the value of the given column for existing rows.
+
+            Overridden here because we skip generating unique access tokens
+            for potentially tons of existing sales orders that won't need it.
+            We do, however, generate it for draft or quotation sales orders.
+        """
+        if column_name != 'access_token':
+            super(SaleOrder, self)._init_column(column_name)
+        else:
+            default_compute = self._fields[column_name].default
+
+            query = """
+                SELECT id FROM "%s"
+                 WHERE "%s" is NULL
+                       AND state in %%s
+            """ % (self._table, column_name)
+            self.env.cr.execute(query, (('draft', 'sent'),))
+            order_ids = self.env.cr.fetchall()
+
+            for [order_id] in order_ids: # order_ids is a list of tuples e.g.: [(1,), (2,), ...]
+                default_value = default_compute(self)
+
+                query = 'UPDATE "%s" SET "%s"=%%s WHERE id = %s' % (
+                    self._table, column_name, order_id)
+                self.env.cr.execute(query, (default_value,))
```

Comment:
Looks like a lot of unnecessary roundtrips to the database.
If the goal is to generate a new random token for each row, how about a single query:

```sql
UPDATE res_users SET access_token = md5(md5(random()::varchar || id::varchar) || clock_timestamp()::varchar)::uuid::varchar
WHERE access_token IS NULL;
```

It does not respect the UUID spec but is sufficiently random for our purposes, I think, and will be very fast even with hundreds of thousands of orders.